### PR TITLE
Non-BP node sync

### PR DESCRIPTION
**Pull request by Jsying**
This PR implements synchronization mechanism, for non-bp nodes.

We need governance state to verify block's signature (randomness),
but in ethereum fast sync mode, eth downloader only downloads the whole
state of pivot block, so we don't have governance state to verify the
downloaded block that is before pivot block if we don't processing
transaction.
To avoid running transactions, dex downloader also downloads the
governance state (merkle proof and storage) at snapshot height of each round,
so that we can verify blocks in fast sync mode.

### CODE REVIEW

File: test/sync_test.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+for i in $(seq 0 3); do
+    datadir=$PWD/Dexon.sync.$i
+    rm -rf $datadir
+    $GDEX --datadir=$datadir init genesis.json
+    $GDEX --testnet --verbosity=4 --gcmode=archive --datadir=$datadir \
+	  --rpc --rpcapi=eth,net,web3,debug --rpcaddr=0.0.0.0 \
+      --port=$((30505 + $i)) \
+      --rpcport=$((8663 + $i *2)) \
+      --wsport=$((8664 + $i * 2))  \
+  	  --ws --wsapi=eth,net,web3,debug --wsaddr=0.0.0.0 \
```
Comment (aitjcize) [inline test/sync_test.sh]: indent

File: core/state/gov.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package state
+
+import (
+	"github.com/dexon-foundation/dexon/common"
+	"github.com/dexon-foundation/dexon/core/types"
+	"github.com/dexon-foundation/dexon/trie"
+)
+
+func GetGovState(statedb *StateDB, header *types.Header,
```
Comment (aitjcize) [inline core/state/gov.go]: can you merge this with `dex/governance.go`? there are some duplicate functions.

File: core/blockchain.go [hunk 3]
```diff
@@ -979,6 +985,9 @@ func (bc *BlockChain) InsertReceiptChain(blockChain types.Blocks, receiptChain [
 			bytes += batch.ValueSize()
 			batch.Reset()
 		}
+		if _, ok := bc.GetRoundHeight(block.Round()); !ok {
```
Comment (aitjcize) [inline core/blockchain.go]: remove this, this is no longer required.

File: core/blockchain.go [hunk 4]
```diff
@@ -1362,6 +1371,205 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
 	return 0, events, coalescedLogs, nil
 }
 
+// InsertChain2 attempts to insert the given batch of blocks in to the canonical
+// chain or, otherwise, create a fork. If an error is returned it will return
+// the index number of the failing block as well an error describing what went
+// wrong.
+//
+// After insertion is done, all accumulated events will be fired.
+func (bc *BlockChain) InsertChain2(chain types.Blocks) (int, error) {
```
Comment (aitjcize) [inline core/blockchain.go]: please help restore `insertChain` to upstream version.

File: core/blockchain.go [hunk 5]
```diff
@@ -1362,6 +1371,205 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
...
+
+			// Only count canonical blocks for GC processing time
+			bc.gcproc += proctime
+
+		case SideStatTy:
+			log.Debug("Inserted forked block", "number", block.Number(), "hash", block.Hash(), "diff", block.Difficulty(), "elapsed",
+				common.PrettyDuration(time.Since(bstart)), "txs", len(block.Transactions()), "gas", block.GasUsed(), "uncles", len(block.Uncles()))
+
+			blockInsertTimer.UpdateSince(bstart)
+			events = append(events, ChainSideEvent{block})
```
Comment (aitjcize) [inline core/blockchain.go]: add panic here, we want to catch this during development.

File: core/governance.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
+package core
```
Comment (aitjcize) [inline core/governance.go]: this is also duplicating functions is `dex/governance.go`. Maybe you should merge it somehow

Comment (aitjcize) [reply core/governance.go]: one solution is to have `dex/governance.go` call the function implemented here.

File: core/blockchain.go (hunk 5 above)
Comment (Jsying) [reply core/blockchain.go]: done

File: core/blockchain.go (hunk 4 above)
Comment (Jsying) [reply core/blockchain.go]: done

File: core/blockchain.go (hunk 3 above)
Comment (Jsying) [reply core/blockchain.go]: done

File: test/sync_test.sh (hunk 1 above)
Comment (Jsying) [reply test/sync_test.sh]: done

File: core/state/gov.go (hunk 2 above)
Comment (Jsying) [reply core/state/gov.go]: refactored

File: core/governance.go (hunk 6 above)
Comment (Jsying) [reply core/governance.go]: refactored

File: core/blockchain.go [hunk 7]
```diff
@@ -1853,6 +2082,24 @@ func (bc *BlockChain) InsertHeaderChain(chain []*types.Header, checkFreq int) (i
...
+	bc.chainmu.Lock()
+	defer bc.chainmu.Unlock()
+
+	bc.wg.Add(1)
+	defer bc.wg.Done()
+
+	whFunc := func(header *types.HeaderWithGovState) error {
+		bc.mu.Lock()
+		defer bc.mu.Unlock()
+		_, err := bc.hc.WriteHeader2(header)
```
Comment (aitjcize) [inline core/blockchain.go]: WriteDexonHeader?

**Pull request merged**